### DISCUSSION TITLE
Experimental Server-Sent Events implementation (package)

### DISCUSSION THREAD

**Discussion by devfrey** [clue/framework-x#68] [Ideas]
Hi,

Since Framework X doesn't yet have an implementation for EventSource / Server-Sent Events, I decided to build an [experimental implementation](https://github.com/devfrey/framework-x-eventsource). It's very much possible I haven't covered all use cases, so looking forward to your feedback. I have no idea if this comes close to what you had in mind, but I hope it helps.

Relevant snippet from the example in the readme:
```php
$app->get('/', new EventSourceHandler($loop, $events = new BufferedEventStream()));

$loop->addPeriodicTimer(1.0, function () use ($events) {
    $events->send((new Event())->data(mt_rand()));
});
```

The object-oriented API for constructing events is fairly straight forward:
```php
$event = (new Event('my-event'))
    ->comment('This is an example event')
    ->data('some-relevant-data');

// In its encoded form:
'
:This is an example event
event: my-event
data: some-relevant-data

';
```

Thanks!

**Comment by clue** [comment]
@devfrey That's awesome! 🎉

I'm indeed currently working on the EventSource documentation here: https://framework-x.clue.engineering/docs/async/streaming/#eventsource

EventSource support is somewhat built-in already because it's literally normal HTTP response streaming, but I agree some documentation and additional helpers to ease getting started makes perfect sense. I've also used variations of https://github.com/clue/php-sse-react in a number of projects and plan to integrate this more closely with Framework X in the future.

What do you think?

**Reply by devfrey** [reply]
My implementation handles some of the necessary stuff to make SSE work smoothly, such as keep-alive, event IDs and the encoding of events. And indeed, the rest is just normal HTTP response streaming.

I guess it's up to you to decide what kind of functionality you want baked into the framework. For now, it could also live as a separate package. It integrates nicely without any modifications. If you don't have any feedback on the implementation, we could consider mentioning it to the framework's Streaming/EventSource documentation.

Happy to hear your thoughts.

**Reply by clue** [reply]
Excellent points!

I still think the documentation should cover the basics and perhaps we can include a link to your implementation for more details?

Happy to discuss this further, perhaps we get to chat at the next meetup? -> https://github.com/clue/framework-x/discussions/5

**Comment by bpolaszek** [comment]
That sounds great! What if events to be written depend on the request (query string, authorization headers, ...)? 
Where would you put that logic?
One year ago I implemented a [Mercure](https://mercure.rocks) server [in ReacPHP](https://github.com/bpolaszek/mercure-php-hub) as a proof-of-concept. 
FrameworkX and your SSE implementation could be good candidates for some refactoring :-)

**Comment by carliedu** [comment]
I would suggest an explaination for what it is for and how it works. In ```clue/php-sse-react/blob/master/README.md``` is no explanation. 
I'm using it because I explored the ```11-chat.php```. It works very well, but honestly I can not explain some things. I have the impression that, its function is to permit sending the same message to more than one client. That's the "buffer" context. Is it only this?
And I have to define myself as a beginner in ReactPHP, even though I've been exploring it for a few months. For me it has been a big paradigm shift, especially debugging. For me, documentation is very important.
